### PR TITLE
Improve handling of frozen VsTest runs

### PR DESCRIPTION
**Pull request by dupdob**
- Add Stryker controlled timeout for VsTest to detect frozen VsTest runs and retry them. Fix test runner frozen after test discovery. May fix: https://github.com/stryker-mutator/stryker-net/issues/1665 , https://github.com/stryker-mutator/stryker-net/issues/1136
- Fix a bug causing the loss of runner due to early disposing
- Fix existing VsTest recycling logic which was not working (**post initial PR**)
- Restore integration tests in the pipeline (**added**)
- Detect when VsTest fails to start and stop the execution altogether (as nothing can be done) (**added**)
- Reduce the computation time of test runs via fixing and improving the existing optimization logic
- Perform minor refactoring (cleaning of the no longer needed `RunAll`method)

### CODE REVIEW

**Comment by eNeRGy164** [comment]
Something is really broken in this branch as almost no test coverage is found anymore.

Running on FluentAssertions with the following config file:
```json
{
    "stryker-config":
    {
        "project": "Src\\FluentAssertions\\FluentAssertions.csproj",
        "test-projects": [
            "Tests\\FluentAssertions.Specs\\FluentAssertions.Specs.csproj",
            "Tests\\FluentAssertions.Equivalency.Specs\\FluentAssertions.Equivalency.Specs.csproj"
        ],
        "reporters": [
            "progress",
            "html"
        ],
        "verbosity": "info",
        "concurrency": 4,
        "mutation-level": "Complete"
    }
}
```

Running with 1.4.2
> [19:25:16 INF] 242   mutants got status CompileError. Reason: Mutant caused compile errors
[19:25:16 INF] **666**   mutants got status NoCoverage.   Reason: Mutant has no test coverage
[19:25:16 INF] 1     mutants got status Ignored.      Reason: Unable to inject back mutations in source code.
[19:25:16 INF] 909   total mutants are skipped for the above mentioned reasons
[19:25:16 INF] **8873**  total mutants will be tested

Running with 1.5.0
> [19:30:30 INF] 242   mutants got status CompileError. Reason: Mutant caused compile errors
[19:30:30 INF] **666**   mutants got status NoCoverage.   Reason: Mutant has no test coverage
[19:30:30 INF] 1     mutants got status Ignored.      Reason: Unable to inject back mutations in source code.
[19:30:30 INF] 909   total mutants are skipped for the above mentioned reasons
[19:30:30 INF] **8873**  total mutants will be tested

Running with this branch
> [12:30:14 INF] 244   mutants got status CompileError. Reason: Mutant caused compile errors
[12:30:14 INF] **9529**  mutants got status NoCoverage.   Reason: Mutant has no test coverage
[12:30:14 INF] 1     mutants got status Ignored.      Reason: Unable to inject back mutations in source code.
[12:30:14 INF] 9774  total mutants are skipped for the above mentioned reasons
[12:30:14 INF] **8**     total mutants will be tested

With only 8 tests, it is fast 😉 but not what I expected

**Comment by dupdob** [comment]
Thanks for the feedback. It is not clear how the changes could have this effect.
I tried to reproduce it with FluentAssertions solution, but It does not build (yet?) on my machine.

**Comment by dupdob** [comment]
Could you provide some logs ? I always checked my PRs against NFluent (4K mutants, 1.7K tests) and it works normally.
I did another check using multiple test projects to have a similar configuration and everything works as expected.

It looks like only one of the test projects was taken into account here. Anyway, I do not think your issue relates to this PR (at this stage)

**Comment by eNeRGy164** [comment]
> Thanks for the feedback. It is not clear how the changes could have this effect. I tried to reproduce it with FluentAssertions solution, but It does not build (yet?) on my machine.

To build everything running `build.ps1/.sh` should be sufficient. But is has some dependencies like the .NET 4.7 SDK, the .NET 6 SDK, including UWP.

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs [hunk 1]
```diff
@@ -508,7 +498,7 @@ private string GenerateRunSettings(int? timeout, bool forMutantTesting, bool for
 
             if (_testFramework.HasFlag(TestFramework.xUnit))
             {
-                settingsForCoverage += "<DisableParallelization>true</DisableParallelization>";
+                settingsForCoverage += "<Disable>true</DisableParallelization>";
```
Comment (eNeRGy164) [inline src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:501]: Here is the bug, the xml tags are inconsistently replaced

Comment (eNeRGy164) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:501]: Looks more promissing:

> [17:19:43 INF] 242   mutants got status CompileError. Reason: Mutant caused compile errors
[17:19:43 INF] **409**   mutants got status NoCoverage.   Reason: Mutant has no test coverage
[17:19:43 INF] 1     mutants got status Ignored.      Reason: Unable to inject back mutations in source code.
[17:19:43 INF] 652   total mutants are skipped for the above mentioned reasons
[17:19:43 INF] **9400**  total mutants will be tested

Comment (eNeRGy164) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:501]: > 100,00% │ Testing mutant 9400 / 9400 │ K 8295 │ S 1020 │ T 85 │ ~0m 00s │
> 
> Killed:   8295
Survived: 1020
Timeout:    85
> 
> [18:38:30 INF] Time Elapsed 01:20:44.0972875
[18:38:30 INF] The final mutation score is 85.43 %

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:501]: Thanks for spotting this. This is indeed some mishap from my part. This is a xUnit specific settings, and NFluent uses NUnit. I understand why I was not able to reproduce it.

[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:501]: Yes, not sure how it happened, but this is a typing stupidity.
I am using this now as an opportunity to improve error handling, so there could be a clear error message when coverage test fails.

**Comment by dupdob** [comment]
I pushed a new commit with the fix (obviously) as well as some added logic to detect similar condition.
This kind of issue should have been detected by integration tests. I will have a look in the coming days to understand why nothing was detected.

File: src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs [hunk 2]
```diff
@@ -173,7 +173,7 @@ private void ReadConfiguration(string configuration)
             }
 
             var nameSpaceNode = node.SelectSingleNode("//Parameters/MutantControl");
-            if (nameSpaceNode != null && nameSpaceNode.Attributes != null)
+            if (nameSpaceNode?.Attributes != null)
```
Comment (eNeRGy164) [inline src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs:176]: If you are changing this line, maybe use the "new' way of comparing

```suggestion
            if (nameSpaceNode?.Attributes is not null)
```

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunnerPool.cs [hunk 3]
```diff
@@ -50,10 +48,7 @@ public TestRunResult TestMultipleMutants(ITimeoutValueCalculator timeoutCalc, IR
             }
         }
 
-        public TestRunResult RunAll(ITimeoutValueCalculator timeoutMs, Mutant activeMutant, TestUpdateHandler update)
-        {
-            return TestMultipleMutants(timeoutMs, activeMutant == null ? null : new List<Mutant> { activeMutant }, update);
-        }
+        public TestRunResult RunAll(ITimeoutValueCalculator timeoutMs, Mutant activeMutant, TestUpdateHandler update) => TestMultipleMutants(timeoutMs, activeMutant == null ? null : new List<Mutant> { activeMutant }, update);
```
Comment (eNeRGy164) [inline src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunnerPool.cs:51]: Wrapping this in one line does not really make it more readable

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs (hunk 1 above)
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs]: as per the other conversation, this problem should have been detected by integration tests. It turns out they have been disabled for several months 😄. Thanks for your contribution that helped us fix two significant issues.

File: src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs (hunk 2 above)
Comment (dupdob) [reply src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs:167]: Not possible: the DataCollector needs to be built against NetStandard 2.0, which forces C# to 7.3 whereas pattern matching features were introduced in C#9.

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunnerPool.cs (hunk 3 above)
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunnerPool.cs]: Thanks for pointing to this. It turns out this method should have been removed altogether; it was a remnant of the early design which did not support testing multiple mutants and was needed while Stryker still supported `dotnet test`  on top of VsTest

File: src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs (hunk 2 above)
Comment (eNeRGy164) [reply src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs:167]: Just add `<LangVersion>latest</LangVersion>` to the csproj?

**Comment by dupdob** [comment]
I discovered that integration testing was no longer effective (incorrect filtering syntax in the pipeline). I have restored them and re established the score base lines (roughly twice as much mutations as before).
Also did some cleanup on the code on an opportunity basis.

**Comment by rouke-broersma** [comment]
> I discovered that integration testing was no longer effective (incorrect filtering syntax in the pipeline). I have restored them and re established the score base lines (roughly twice as much mutations as before). Also did some cleanup on the code on an opportunity basis.

That is... 😶

**Comment by eNeRGy164** [comment]
Isn't this becoming a bit of scope creep?

File: src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs [hunk 4]
```diff
@@ -233,22 +224,16 @@ private void PublishCoverageData(TestCaseEndArgs testCaseEndArgs)
             }
 
             _dataSink.SendData(testCaseEndArgs.DataCollectionContext, PropertyName, coverData);
-            if (_mutationCoveredOutsideTests.Count > 0)
-            {
-                // report any mutations covered before this test executed
-                _dataSink.SendData(testCaseEndArgs.DataCollectionContext, OutOfTestsPropertyName, string.Join(",", _mutationCoveredOutsideTests));
-                _mutationCoveredOutsideTests.Clear();
-            }
+            if (_mutationCoveredOutsideTests.Count <= 0) return;
```
Comment (rouke-broersma) [inline src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs:227]:
```suggestion
            if (_mutationCoveredOutsideTests.Count <= 0) { return; }
```

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunnerPool.cs [hunk 5]
```diff
@@ -26,14 +26,12 @@ public VsTestRunnerPool(StrykerOptions options, ProjectInfo projectInfo)
...
+            var runner = new VsTestRunner(options, projectInfo, null, new TestSet(), 0, helper: _helper);
             (_vsTests, _tests) = runner.DiscoverTests(null);
             _availableRunners.Add(runner);
 
-            Parallel.For(1, options.Concurrency, (i, loopState) =>
-            {
-                _availableRunners.Add(new VsTestRunner(options, projectInfo, _vsTests, _tests, helper: _helper));
-            });
+            Parallel.For(1, options.Concurrency,
+                (i, _) => _availableRunners.Add(new VsTestRunner(options, projectInfo, _vsTests, _tests, i, helper: _helper)));
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunnerPool.cs:34]:
```suggestion
            Parallel.For(1, options.Concurrency, (i, _) => 
                 _availableRunners.Add(new VsTestRunner(options, projectInfo, _vsTests, _tests, i, helper: _helper)));
```

**Comment by eNeRGy164** [comment]
> What do you feel is scope creep here ? restoring the integration tests is part of the CI and is (obviously) top priority, especially since we could have released a faulty version if you have not tried it on your own (thanks again, btw)

Yes, I understand the priority, but it's not introduced by this change, and like you discovered, already present for a while.
But now this PR is no longer only about VSTest anymore. Meanwhile people having issues with time-outs will be held back on a fix for something that's not affecting them.

Personally I would start fixing the integration tests in a separate PR, resulting also in a separate release note, etc.
But maybe that's just how I'm used to do PR's.

It's more of a warning, in the end, I'm just a consumer, not a maintainer, just trying to be helpful that you don't get a big ball of mud in changes that are hard to release until the next thing is fixed.

**Comment by dupdob** [comment]
PR is ready for merge (fixed the problem with the FullFramework test project using an obsolete framework version).

File: src/Stryker.Core/Stryker.Core/CoverageAnalysis/CoverageAnalyser.cs [hunk 6]
```diff
@@ -72,6 +71,19 @@ private void SetCoveringTests(IReadOnlyCollection<Mutant> mutantsToScan)
                 {
                     mutant.CoveringTests = TestsGuidList.EveryTest();
                 }
+                if (mutant.CoveringTests.IsEveryTest)
+                {
+                    _logger.LogDebug($"Mutant {mutant.Id} will be tested agaisnt all tests.");
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/CoverageAnalysis/CoverageAnalyser.cs:76]:
```suggestion
                    _logger.LogDebug($"Mutant {mutant.Id} will be tested against all tests.");
```

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs [hunk 7]
```diff
@@ -186,9 +182,11 @@ void HandleUpdate(IRunResults handler)
                 _aborted = true;
             }
 
-            _logger.LogDebug("Using {0} ms as testrun timeout", timeOutMs.ToString() ?? "no");
+            if (timeOutMs.HasValue)
+                _logger.LogDebug($"{RunnerId}: Using {timeOutMs} ms as test run timeout");
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:186]:
```suggestion
            if (timeOutMs.HasValue)
            {
                _logger.LogDebug($"{RunnerId}: Using {timeOutMs} ms as test run timeout");
            }
```

**Pull request merged**